Update `minSdkVersion` (from `21` to `24`)

ParaskP7 opened the pull request:
This PR upgrades FluxC to `minSdkVersion` to `24`. This has been unblocked because:
- `WPAndroid` was already on `minSdkVersion = 24`, for some time now (see [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15135)).
- `WCAndroid` has been recently upgraded to `minSdkVersion = 24` as well (see [here](https://github.com/woocommerce/woocommerce-android/pull/7643)).

-----

Warning (⚠️): This update is currently blocked by the common upgrade of `targetSdkVersion` to `31` (see [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2563)).

-----

As part of this `minSdkVersion = 24` upgrade the below changes were also applied in order to fix any additional warnings that this change brought-up:

Warnings Resolution List:

1. [Resolve obsolete sdk int lint warning for well sql config.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/986fc83303eb5472d38ff18a25982c17ceff64f5)
2. [Resolve obsolete sdk int lint warning for open jdk cookie mng.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/3400d7b0a8a6e33a3ca2c943a47107376be896c2)
3. [Resolve obsolete sdk int lint warning for leaderboard prd item.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/ea8f79dc9f1bd4a7977cbd0386f21f7caf4e0a2c)
4. [Resolve obsolete sdk int lint warning for mocked statck test.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/ef933c2349c6d66cc545aba6f03d0dfee1271694)

-----

PS: @hichamboushaba @ovitrif I added you as the main reviewers, that is, in addition to the @wordpress-mobile/apps-infrastructure team itself, but randomly, since I just wanted someone from both, the `WordPress` and `Woo Commerce` teams, to sign-off on that change for WPAndroid and WCAndroid alike.

-----

### Testing instructions

- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough.
- However, if you want to be thorough about reviewing this change, you could:
    - Quickly smoke test the `FluxC Example` app and see if it works as expected.
    - Try and install the `FluxC Example` app on Android 5 (`API 21/22`) and Android 6 (`API 23`) devices and verify that you can't (see [API Levels](https://apilevels.com/)).

-----

### Merge instructions

- [x] Wait for #2563 to be approved and merged to `trunk`.
- [x] Make sure this PR's base is automatically updated to `trunk`.
- [x] Update PR with latest `trunk`.
- [x] Remove `[PR] Not Ready For Merge]` label.
- [ ] Merge PR to `trunk`.

ParaskP7 committed:
Build: Update min sdk version to 24

Although the 'assemble' and 'lint' task run successfully, a quick
analysis showed that updating to 'minSdkVersion = 24' added/removed the
below extra warnings:

Lint Warnings:
- +2 x ObsoleteSdkInt (plugins:woocommerce)

All these newly created warnings will be dealt with in subsequent
commits. Also, any existing such warnings will be resolved too.

ParaskP7 committed:
Analysis: Resolve obsolete sdk int lint warning for well sql config

Warning Message: "Unnecessary; SDK_INT is always >= 24"

This 'VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN' check was, and
is, no longer necessary as the 'minSdkVersion' is now '24'.

ParaskP7 committed:
Analysis: Resolve obsolete sdk int lint warning for open jdk cookie mng

Warning Message: "Unnecessary; SDK_INT is always >= 24"

This 'VERSION.SDK_INT >= Build.VERSION_CODES.N' check is no longer
necessary as the 'minSdkVersion' is now '24'. As such, the if condition
has been simplified.

ParaskP7 committed:
Analysis: Resolve obsolete sdk int lint warning for leaderboard prd item

Warning Message: "Unnecessary; SDK_INT is always >= 24"

Both, this '@RequiresApi(VERSION_CODES.N)' annotation and
'VERSION.SDK_INT >= Build.VERSION_CODES.N' check, are no longer
necessary as the 'minSdkVersion' is now '24'.

ParaskP7 committed:
Analysis: Resolve obsolete sdk int lint warning for mocked statck test

Warning Message: "Unnecessary; SDK_INT is always >= 24"

This 'assumeTrue(Build.VERSION.SDK_INT >= 23)' test assertion is no
longer necessary as the 'minSdkVersion' is now '24'.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/leaderboards/LeaderboardProductItem.kt:
```diff
@@ -175,11 +173,9 @@ class LeaderboardProductItem(
-            ?.takeIf { Build.VERSION.SDK_INT >= Build.VERSION_CODES.N }?.let {
-                Html.fromHtml(source, Html.FROM_HTML_MODE_LEGACY)
-            } ?: Html.fromHtml(source)
+    private fun fromHtmlWithSafeApiCall(source: String?) = source?.let {
+        Html.fromHtml(source, Html.FROM_HTML_MODE_LEGACY)
+    } ?: Html.fromHtml(source)
```
I think we can simplify this function by removing the call to the parameterless `fromHtml`, as it's equivalent to the other one using `FROM_HTML_MODE_LEGACY`: https://developer.android.com/reference/android/text/Html#fromHtml(java.lang.String). This was necessary before because of the `takeIf`, but not anymore I think.

ParaskP7 replied:
Great, thanks for the comment @hichamboushaba ! I agree, the cleaner the better, I'll apply this change and let you know as soon as I push its commit. 🙇

ParaskP7 replied:
👋 @hichamboushaba !

This is now done, see ba0b9060f502d239d7a235ea8cbe8bc4667642c5 and 8637d1cb309b769edced2c03af6f77c5592e1449.

You will note that I went a bit further and tried to also resolve the `Deprecated` warning, but I only could go as far as doing so for 1/3 of those. You will find further detail in the commit description itself. Let me know what you think and thank you! 🙇

hichamboushaba replied:
Thanks @ParaskP7 for making the changes 🙏 , I left some comments below.
All of this is minor, so no need for spending too much time on it 🙂.

hichamboushaba reviewed (approved):
Looks good, thanks @ParaskP7 for working on this.
I left one comment, but it's not blocking, so I'm pre-approving.

ParaskP7 commented:
Thank you so much for reviewing and testing this PR @hichamboushaba , you rock! 🙏

ParaskP7 committed:
Refactor: Replace from html with safe api call with html from html call

This 'fromHtmlWithSafeApiCall(...)' function served a purpose only when
it was combined with `takeIf`. However, now that this was removed due to
the 'minSdkVersion = 24' upgrade, this whole function is no longer
necessary and can be replaced with a direct 'Html.fromHtml(...)' call.

ParaskP7 committed:
Analysis: Resolve new api lint warnings for html from html static call

Replacing 'Html.fromHtml(html)' with the
'HtmlCompat.fromHtml(html, HtmlCompat.FROM_HTML_MODE_LEGACY)' compat
related version is the recommended action to resolve this kind of Lint
warnings.

However, note that this was only possible for 'priceAmountHtmlTag'.
Doing the same for the other two, 'priceAmountHtmlTag' and
'itemHtmlTag', isn't possible as those fields can end up being nullable,
that is, due to the fact that the 'LeaderboardItemRow.display'
constructor argument is nullable as well. As such, and because
'HtmlCompat.fromHtml(html, HtmlCompat.FROM_HTML_MODE_LEGACY)' accepts a
non-null String 'source', it is not possible to resolve these two Lint
warnings.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/leaderboards/LeaderboardProductItem.kt:
```diff
@@ -84,7 +85,7 @@ class LeaderboardProductItem(
                 ?.filter { it.contains("&#") }
                 ?.reduce { total, new -> "$total$new" }
-                ?.run { Html.fromHtml(this) }
+                ?.run { HtmlCompat.fromHtml(this, HtmlCompat.FROM_HTML_MODE_LEGACY) }
```
np, you don't have to change it.
`HtmlCompat` is useful only for apps with minSdk less than 24, for us, we can directly call the method that was introduced in API 24 https://developer.android.com/reference/android/text/Html#fromHtml(java.lang.String,%20int)

ParaskP7 replied:
Ah, thank you @hichamboushaba , you are totally right, I somehow got afraid of the fact that the `String source` is neither marked `@NonNull` nor `@Nullable` on the `Html.java` side and thus didn't want to make this change, but that doesn't matter, you have to trust that it works for `@Nullable` just like that as it anyways defaults to using the same mechanism underneath.

Thus, done, this update seems safe and apologies for the trouble with this review (😅): 5acc19a9206c17ed7d07498cd73f4cdb31134119

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/leaderboards/LeaderboardProductItem.kt:
```diff
@@ -150,7 +150,7 @@ class LeaderboardProductItem(
      */
     private val link by lazy {
-        fromHtmlWithSafeApiCall(itemHtmlTag)
+        Html.fromHtml(itemHtmlTag)
```
For using the new API instead of the deprecated one, WDYT about something like this?
```kotlin
itemHtmlTag?.let { Html.fromHtml(it, Html.FROM_HTML_MODE_LEGACY) }
                  ?.run { this as? SpannableStringBuilder }
                 ...
```

ParaskP7 replied:
👋 @hichamboushaba !

I think maybe [this change](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2564#discussion_r1018976760) is enough for now, wdyt? 🤔

PS: I really don't want to change any of the functionality, even by that slight bit, not that it will change much, but you know, you can never be too careful, and then you need to really test that, even if you change that little bit.

hichamboushaba replied:
Good point, and I think we are safe, when the argument is `null`, we were already calling `Html#fromHtml(String)` which proxies the call to the same function we are using now.

ParaskP7 replied:
Exactly, once more, thank YOU Hicham ! 🙇

ParaskP7 committed:
Refactor: Use html from html with from html mode legacy flag static call

As correctly noted 'HtmlCompat' is only useful for apps with
`minSdkVersion` less than '24', thus it is now unneeded and can be
completely replaced with the direct usage of 'Html'.

See discussion here: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/pull/2564#discussion_r1018958379

ovitrif reviewed (approved):
Changes LGTM 🎉 

Though I wonder with regards to the existing discussion here, in WPAndroid we always use `HtmlCompat`, yet `minSdkVersion` is 24, isn't it then in fact not necessary? 😅

ParaskP7 commented:
Thank you for the review and testing @ovitrif ! 🙇 ❤️ 🚀 

> Though I wonder with regards to the existing discussion here, in WPAndroid we always use HtmlCompat, yet minSdkVersion is 24, isn't it then in fact not necessary? 😅

Yeap, yeap, this is indeed the case, I just copy-pasted my solution from WCAndroid into WPAndroid, and since I used `HtmlCompat` there, I used it blindly as is on WPAndroid as well. But this is indeed not necessary and we could optimize on it. 💯 

PS: Feel free to do a quick PR on it if you like, else I'll get to that at some point myself. 👍

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/update-min-sdk-version-to-24

 Conflicts:
	settings.gradle

ParaskP7 committed:
Build: Update min sdk version to 24

This is done once more, due to the fact that a merge conflict resulting
in that update being lost.

ParaskP7 committed:
Analysis: Remove unnecessary expired target sdk version lint exclusion

Since the 'targetSdkVersion' has been now update to point to '31',
disabling this 'ExpiredTargetSdkVersion' Lint error is no longer
necessary.
